### PR TITLE
Line oriented fypp parser

### PR DESCRIPTION
**Pull request by masatake**
(Close #1810).

#1810 is token oriented Fypp parser.
This one is line oriented.
The advantage of this parser can run Fortran parser (or anything parser specified with guest parameter) as a guest parser.

Currently all if/elseif/else/endif blocks are masked when running Fortran parser.
This is different from what CPreProcessor parser does. CPreProcessor doesn't mask the first if block in if/else block chain. Fypp parser should do the same.

### CODE REVIEW

**Comment by masatake** [comment]
@p-vitt, could you try this one?

**Comment by p-vitt** [comment]
Sure, I'll have a look at it. Looks interesting, even though I can't grasp it right away. Thanks a lot!

**Comment by p-vitt** [comment]
I gave it a try, and after some changes to my configured ctags command line, it works. I somehow had .fy added to the Fortran langmap. That prevented those files from being parsed by the Fypp parser. After removing that, it worked pretty fine.

But I have one more question. We use .fy for Fypp files containing only Fypp code. We use .fpp for Fortran source code files that need to be preprocessed by Fypp. So I added them to the Fortran langmap. But those .fpp files could also contain Fypp code like macro definitions. How to get them also parsed by the ctags Fypp parser? Is it sufficient to also add them to the Fypp langmap? Or is this somehow related to subparsers?

**Comment by masatake** [comment]
Remove .fpp from the Fortran's langmap.
Then add .fpp to the Fypp's langmap:

```
--map-Fortran=-.fpp --map-Fypp=+.fpp
```

BTW, is using .fpp extension for Fypp popular?
In that case, I should add the rule for .fpp to ctags itself.

**Comment by masatake** [comment]
> Thanks. Adding fpp to Fypp's langmap, but not not Fortran's, works.
... I wonder why it doesn't work.

I will show an example session:

```
[yamato@master]~/var/ctags-github% cat input.fpp
cat input.fpp
#:def debug_code(code)
  #:if DEBUG > 0
    $:code
  #:endif
#:enddef debug_code

program helloworld
print *, "Hello, world."
end program helloworld
[yamato@master]~/var/ctags-github% u-ctags --options=NONE --extras=+g --map-Fypp=+.fpp -o - input.fpp
<s=NONE --extras=+g --map-Fypp=+.fpp -o - input.fpp
u-ctags: Notice: No options will be read from files or environment
debug_code	input.fpp	/^#:def debug_code(code)$/;"	m
helloworld	input.fpp	/^program helloworld$/;"	p
```

> Using fpp extension is quite common for Fortran files that need to be preprocessed. But the extension doesn't tell which preprocessor to use. There are, besides Fypp, also Coco and the C preprocessor. And I bet there are even more. So linking fpp to Fypp isn't IMHO correct.

I see. Thank you.

**Comment by p-vitt** [comment]
Sorry, my last messages seems to be a little incomprehensible. It also works for me. So, everything's fine. Thanks again.

**Comment by masatake** [comment]
Oh, I see.

**Comment by masatake** [comment]
I will push the updated version after merging #1828.

**Comment by masatake** [comment]
The updated code is pushed.

How the lines are processed before calling Fortran parser as a guest parser can be traced with ``--_trace=Fypp`` option.

```
./configure --enable-debugging
make
./ctags --_trace=Fypp ...
```

File: parsers/fypp.c [hunk 1]
```diff
@@ -0,0 +1,299 @@
+/*
+ *   Copyright (c) 2018 Masatake YAMATO
```
[the remark this answers is not in the document]
Comment (masatake) [reply parsers/fypp.c]: Fixed.

**Pull request merged**